pkg/cloud_provider/user_tags.go

```diff
@@ -0,0 +1,523 @@
+func (c *tagValuesClient) close() {
+	if err := c.Close(); err != nil {
+		klog.V(1).ErrorS(err, "failed to close tag value client connection")
+	}
+}
+
+// validateTagExist checks whether the tag exist using the tag's Namespaced name.
+func (c *tagValuesClient) validateTagExist(ctx context.Context, name string) error {
+	if _, err := c.GetNamespacedTagValue(ctx, &rscmgrpb.GetNamespacedTagValueRequest{
+		Name: name,
+	}); err != nil {
+		return fmt.Errorf("failed to fetch %s tag: %w", name, err)
+	}
+	return nil
+}
+
+// close the tag binding client connection to API server.
+func (c *tagBindingsClient) close() {
+	if err := c.Close(); err != nil {
+		klog.V(1).ErrorS(err, "failed to close tag binding client connection")
+	}
+}
+
+// getEffectiveTagList returns the list of tags attached to the resource,
+// which were inherited or attached to the resource directly.
+func (c *tagBindingsClient) getEffectiveTagList(ctx context.Context, rscName string) resourceTags {
+	effectiveTags := make(resourceTags)
+	bindings := c.listEffectiveTags(ctx, rscName)
+	// a resource can have a maximum of {gcpMaxTagsPerResource} tags attached to it.
+	// Will iterate for {gcpMaxTagsPerResource} times in the worst case scenario, if
+	// none of the break conditions are met. Should the {gcpMaxTagsPerResource} be
+	// increased in the future, it should not create an issue, since this is an optimization
+	// attempt to reduce the number of tag write calls by skipping already existing tags,
+	// since tag write operation has quota restriction.
+	for i := 0; i < maxTagsPerResource; i++ {
+		binding, err := bindings.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil || binding == nil {
+			// on encountering any error will continue adding refined tags
+			// which will still have all the user provided tags except for
+			// the removed already existing tags processed until this point
+			// and would end up adding tags which already exist, and error
+			// handling is present for the same.
+			klog.V(5).Infof("failed to list effective tags on the %s resource: %v: %v", rscName, binding, err)
+			break
+		}
+		effectiveTags[binding.NamespacedTagValue] = resourceTagsValueFiller
+	}
+	return effectiveTags
+}
+
+func (c *tagBindingsClient) listEffectiveTags(ctx context.Context, name string) *rscmgr.EffectiveTagIterator {
+	return c.ListEffectiveTags(ctx, &rscmgrpb.ListEffectiveTagsRequest{
+		Parent: name,
+	})
+}
+
+// getTagsToBind returns the filtered list of tags after removing the tags which
+// already exist on the resource.
+func (c *tagBindingsClient) getTagsToBind(ctx context.Context, rscName string, tags resourceTags) (resourceTags, error) {
+	effectiveTags := c.getEffectiveTagList(ctx, rscName)
+	refinedTags := make(resourceTags)
+
+	for tag := range tags {
+		if _, exist := effectiveTags[tag]; !exist {
+			refinedTags[tag] = resourceTagsValueFiller
+		}
+	}
+	klog.V(1).Infof("refined list of resource tags: %v", refinedTags)
```

Comment:
Does this need to be V(1)? It seems like this could result in a lot of log spam, and wouldn't provide that much context for the user to debug